examples/wmt/train.py

```diff
@@ -0,0 +1,638 @@
+   label_smoothing: label smoothing constant, used to determine the on and
+     off values.
+
+  Returns:
+    Tuple of scalar loss and batch normalizing factor.
+  """
+  if logits.ndim != targets.ndim + 1:
+    raise ValueError('Incorrect shapes. Got shape %s logits and %s targets' %
+                     (str(logits.shape), str(targets.shape)))
+  vocab_size = logits.shape[-1]
+  confidence = 1.0 - label_smoothing
+  low_confidence = (1.0 - confidence) / (vocab_size - 1)
+  normalizing_constant = -(
+      confidence * jnp.log(confidence) + (vocab_size - 1) *
+      low_confidence * jnp.log(low_confidence + 1e-20))
+  soft_targets = common_utils.onehot(
+      targets, vocab_size, on_value=confidence, off_value=low_confidence)
+
+  loss = -jnp.sum(soft_targets * nn.log_softmax(logits), axis=-1)
+  loss = loss - normalizing_constant
+
+  normalizing_factor = jnp.prod(targets.shape)
+  if weights is not None:
+    loss = loss * weights
+    normalizing_factor = weights.sum()
+
+  return loss.sum(), normalizing_factor
+
+
+def compute_weighted_accuracy(logits, targets, weights=None):
+  """Compute weighted accuracy for log probs and targets.
+
+  Args:
+   logits: [batch, length, num_classes] float array.
+   targets: categorical targets [batch, length] int array.
+   weights: None or array of shape [batch, length]
+
+  Returns:
+    Tuple of scalar loss and batch normalizing factor.
+  """
+  if logits.ndim != targets.ndim + 1:
+    raise ValueError('Incorrect shapes. Got shape %s logits and %s targets' %
+                     (str(logits.shape), str(targets.shape)))
+  loss = jnp.equal(jnp.argmax(logits, axis=-1), targets)
+  normalizing_factor = jnp.prod(logits.shape[:-1])
+  if weights is not None:
+    loss = loss * weights
+    normalizing_factor = weights.sum()
+
+  return loss.sum(), normalizing_factor
+
+
+def compute_metrics(logits, labels, weights, label_smoothing=0.0):
+  """Compute summary metrics."""
+  loss, weight_sum = compute_weighted_cross_entropy(logits, labels, weights,
+                                                    label_smoothing)
+  acc, _ = compute_weighted_accuracy(logits, labels, weights)
+  metrics = {
+      'loss': loss,
+      'accuracy': acc,
+      'denominator': weight_sum,
+  }
+  metrics = jax.lax.psum(metrics, axis_name='batch')
+  return metrics
+
+
+def train_step(optimizer,
+               batch,
+               learning_rate_fn,
+               label_smoothing=0.0,
+               use_bfloat16=False,
+               dropout_rng=None):
+  """Perform a single training step."""
+  train_keys = ['inputs', 'targets',
+                'inputs_position', 'targets_position',
+                'inputs_segmentation', 'targets_segmentation']
+  (inputs, targets,
+   inputs_positions, targets_positions,
+   inputs_segmentation, targets_segmentation) = [
+       batch.get(k, None) for k in train_keys]
+
+  weights = jnp.where(targets > 0, 1, 0).astype(jnp.float32)
```

Comment:
Why do we use a weighted loss if the weights have no effect anyway? It seems we multiply the target by 0 if it was 0, and by 1 if it wasn't zero. So it seems like it won't change `target` at all.

Reply:
I'm not sure I understand - targets get turned into one-hot vectors (or label-smoothed almost one-hot vectors) to calc. cross-entropy, so you definitely still want to mask out padding tokens and keep the weight-sums as well to correctly calculate cross-entropy as well as to aggregate metrics across steps.

Reply:
Ah sorry, I misunderstood.